Src/WindowsServer/WindowsServer.Shared/AzureInstanceMetadataTelemetryModule.cs

```diff
@@ -0,0 +1,56 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation;
+
+    /// <summary>
+    /// A telemetry module that adds Azure instance metadata context information to the heartbeat, if it is available.
+    /// </summary>
+    public sealed class AzureInstanceMetadataTelemetryModule : ITelemetryModule
+    {
+        private bool isInitialized = false;
+        private object lockObject = new object();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureInstanceMetadataTelemetryModule" /> class.
+        /// Creates a heartbeat property collector that obtains and inserts data from the Azure Instance
+        /// Metadata service if it is present and available to the currently running process. If it is not
+        /// present no added IMS data is added to the heartbeat.
+        /// </summary>
+        /// <param name="unused">Unused parameter for this TelemetryModule.</param>
+        public void Initialize(TelemetryConfiguration unused)
+        {
+            // Core SDK creates 1 instance of a module but calls Initialize multiple times
+            if (!this.isInitialized)
+            {
+                lock (this.lockObject)
+                {
+                    if (!this.isInitialized)
+                    {
+                        var telemetryModules = TelemetryModules.Instance;
+
+                        foreach (var module in telemetryModules.Modules)
+                        {
+                            if (module is IHeartbeatPropertyManager hbeatManager)
+                            {
+                                // start off the heartbeat property collection process, but don't wait for it nor report
+                                // any status from here, fire and forget. The thread running the collection will report 
+                                // to the core event log.
+                                var heartbeatProperties = new AzureComputeMetadataHeartbeatPropertyProvider();
+                                Task.Factory.StartNew(
```

Comment:
Do we need to make sure it does not throw, so it won't break the app? I see that we catch web request exception down the line - should we have a global handling as well right within this task to keep the app running in case of any other issues?

Reply:
Agreed. I've added a try...catch and an event log entry for this.

Reply:
And yes, I've opened issues for all outstanding requests from this PR and marked them as due for beta2 milestone.